docs/contributing/testing-guidelines/unit-testing.md

```diff
@@ -0,0 +1,168 @@
+framework is also used to analyze the test results.
+
+`ament_cmake` provides several convenience functions to make it easy to register tests in a
+CMake-based package and to ensure that JUnit-compatible result files are generated. It currently
+supports a few different testing frameworks like `pytest`, `gtest`, and `gmock`.
+
+In order to prevent tests running in parallel from interfering with each other when publishing and subscribing to ROS topics,
+it is recommended to use commands from [`ament_cmake_ros`](https://github.com/ros2/ament_cmake_ros/tree/master/ament_cmake_ros/cmake) to run tests in isolation.
+
+See below for an example of using `ament_add_ros_isolated_gtest` with `colcon test`.
+All other tests follow a similar pattern.
+
+## Create a unit test with gtest
+
+In `my_cool_pkg/test`, create the `gtest` code file `test_my_cool_pkg.cpp`:
+
+```cpp
+#include "gtest/gtest.h"
+#include "my_cool_pkg/my_cool_pkg.hpp"
+TEST(TestMyCoolPkg, TestHello) {
+  EXPECT_EQ(my_cool_pkg::print_hello(), 0);
+}
+```
+
+For more examples of `gtest` features, see the
+[gtest repo](https://github.com/google/googletest).
+
+In `package.xml`, add the following line:
+
+```xml
+<test_depend>ament_cmake_ros</test_depend>
+```
+
+Next add an entry under `BUILD_TESTING` in the `CMakeLists.txt` to compile the test
+source files:
+
+```cmake
+if(BUILD_TESTING)
+  find_package(ament_lint_auto REQUIRED)
+  ament_lint_auto_find_test_dependencies()
+
+  ament_add_ros_isolated_gtest(${TEST_MY_COOL_PKG} test/test_my_cool_pkg.cpp)
+
+  target_link_libraries(${TEST_MY_COOL_PKG} ${PROJECT_NAME})
+...
+endif()
+```
+
+This automatically links the test with the default main function provided by `gtest`. The code under test is usually in a different CMake target (`${PROJECT_NAME}` in the example) and its shared object for linking need to be added.
+
+To register a new `gtest` item, wrap the test code with the macro `TEST ()`. `TEST ()`
+is a predefined macro that helps generate the final test code, and also registers
+a `gtest` item to be available for execution.
+The test case name should be in CamelCase, since gtest inserts an underscore between the fixture name and the class case name when creating the test executable.
+
+`gtest/gtest.h` also contains predefined macros of `gtest` like `ASSERT_TRUE(condition)`,
+`ASSERT_FALSE(condition)`, `ASSERT_EQ(val1,val2)`, `ASSERT_STREQ(str1,str2)`,
+`EXPECT_EQ()`, etc. `ASSERT_*` will abort the test if the condition is not
+satisfied, while `EXPECT_*` will mark the test as failed but continue to next test
+condition. More information about `gtest` can be found in the
+[gtest repo](https://github.com/google/googletest).
+
+In the demo `CMakeLists.txt`, `ament_add_ros_isolated_gtest` is a predefined macro in `ament_cmake_ros`
+that helps simplify adding `gtest` code. Details can be viewed in
+[ament_add_gtest.cmake](https://github.com/ros2/ament_cmake_ros/tree/master/ament_cmake_ros/cmake).
+
+## Build test
+
```

Comment:
Like this.

```suggestion

<!-- cspell:ignore Testfile -->

```

Reply:
Does this line need to be anywhere specific, or can I just add it to the end of the file?

Reply:
It seems anywhere including the end of the file is acceptable.